Helper/Cart.php

```diff
@@ -483,7 +483,49 @@ protected function getCartCacheIdentifier($cart)
         // display_id is always different for every new cart / immutable quote
         // unset it in the cache identifier so the rest of the data can be matched
         unset ($cart['display_id']);
-        return md5(json_encode($cart));
+        $identifier  = json_encode($cart);
+        // extend cache identifier with custom address fields
+        $identifier .= $this->convertCustomAddressFieldsToCacheIdentifier($this->getLastImmutableQuote());
+
+        return md5($identifier);
+    }
+
+    /**
+     * Create cache identifier string from custom address fields
+     *
+     * @param Quote $quote
+     * @return string
+     */
+    public function convertCustomAddressFieldsToCacheIdentifier($quote) {
+
+        // get custom address fields to be included in cache key
+        $prefetchAddressFields = explode(
+            ',', $this->configHelper->getPrefetchAddressFields($quote->getStoreId())
+        );
+        // trim values and filter out empty strings
+        $prefetchAddressFields = array_filter(array_map('trim', $prefetchAddressFields));
+        // convert to PascalCase
+        $prefetchAddressFields = array_map(
+            function ($el) {
+                return str_replace('_', '', ucwords($el, '_'));
+            },
+            $prefetchAddressFields
+        );
```

Comment:
Maybe better to extract these three variables `$prefetchAddressFields` and operations with them to the separate method.
After that we have good ability to cover it with unit tests.

Reply:
@YevhenShBolt Done